### DISCUSSION TITLE
Bug in Querying mongoDB in getStaticProps ?

### DISCUSSION THREAD

**Discussion by oshanfernando** [vercel/next.js#13457] [Help]
Although the documentation states that we can write direct database queries in the **getStaticProps** method, it seems like querying the MongoDB from a field other than the **_id** does not work.

`await client.db('blogdb').collection('posts').findOne({'urlname': params.meta}).toArray()` throws the the following error:  
```
./node_modules/mongodb/lib/url_parser.js
Module not found: Can't resolve 'dns'
```

But if I query from the id, it works
`await client.db('blogdb').collection('posts').find(ObjectID(params.meta)).toArray()`

If anyone came across this issue, please suggest a workaround for this, other than writing a separate backend api to fetch the data.

**Comment by lukaszsobek** [comment]
The thing that might be happening is that the code is executed client side. If there is an instance of MongoClient executed outside of `getStaticProps` and/or `getStaticPaths` it would get called on the client side as well and lead to the error you mentioned (since the client side has no dns). Could you post how the code imports MongoClient (I assume that is what is being used) and the way it is used?

**Reply by oshanfernando** [reply]
@lukaszsobek  you're right, I'm using the MongoClient and this is how i imported the client.
`import { MongoClient } from 'mongodb'`

And I'm creating a new Client inside the getStaticProps as such
`const client = new MongoClient(MONGO_CONNECTION_STRING, {
        useUnifiedTopology: true,
    });`

**Reply by lukaszsobek** [reply]
Seems good to me (tested on 9.3.4). You might dig further into the issue by creating a page like this:
```
import { MongoClient } from "mongodb";

const test = (props) => {
    return <h1>Hello</h1>;
};

export default test;

export const getStaticProps = () => {
    const client = new MongoClient();
    return { props: {} };
};
```
If that works the module is only being imported server side and the treeshaking did its job. Yay.

Now try commenting out the `const client = new MongoClient();` and it should throw an error similar to the one you mentioned above. (btw. It's interesting that a module that is imported but not referenced is parsed client side where it might be better off being ignored altogether.) 

Hope you can work it out from here, because if the test above works the error must be somewhere in the remaining code only you have access to, maybe a stray import, a function called from outside of getStaticProps that references urlParser etc.

**Reply by lukaszsobek** [reply]
btw. below is a sample of code I used to test the getStaticProps:
```
    const client = new MongoClient(
        `mongodb+srv://${process.env.MONGO_USER}:${process.env.MONGO_PW}@${process.env.MONGO_URI}`,
        {
            useNewUrlParser: true,
            useUnifiedTopology: true,
        }
    );

    if (!client.isConnected()) await client.connect();
    const db = client.db("site-test");

    const doc = await db.collection("items").findOne({ brandId: brandId });
```

**Reply by oshanfernando** [reply]
Thank you @lukaszsobek I tried the your code and it worked perfectly fine in another page. I took the code as it is and used it in my actual page but then it throws that error if i query from another field other that the id.  :( 

So in the below code, if i query like `collection('posts').find(ObjectID(params.postid))` it works fine.
I'd be really glad if you could let me know if something is off in these 2 functions.
Sorry i cant paste the code, it gets messed up :/
[Please view this gist](https://gist.github.com/oshanfernando/528d9f0d4d3b1d7a322edd1ebfbc4c9d)
